Sign up done inshallah

aabdoo23 opened the pull request:
#### PR Classification
New feature: Introduced MediatR for handling student creation and set up the database schema.

#### PR Summary
This PR introduces MediatR for handling student creation, sets up the database schema, and updates project configurations.
- Added `StudentsController` in `StudentController.cs` using MediatR for student creation.
- Created initial migration files to set up the database schema.
- Updated `Program.cs` to register new services for persistence, infrastructure, and MediatR.
- Added repository and service interfaces and implementations for `Admin`, `Major`, `Moderator`, and `Student` entities.
- Configured `ForumsDbContext` and added `DbContextRegistrationService` for PostgreSQL database integration.

aabdoo23 committed:
Bismillah

aabdoo23 committed:
first commit, entities not done yet

Add new entities, validators, and project references

Updated SchedsForums.Application.csproj and SchedsForums.Domain.csproj to include new package and project references. Cleaned up ConfigureServices.cs by removing unused using directives. Removed IBaseEntity.cs and IBaseRepository.cs from SchedsForums.Domain. Added new entity classes (Admin, Major, Moderator, Student) and base classes (BaseEntity, BaseUser). Introduced new validators and services, including ConfigureValidators.cs and PasswordService.cs. Created SchedsForums.Interface project with new interfaces and project references.

aabdoo23 committed:
Add persistence services and refactor domain models for the user sign up, next step is the repository, commands and queries, and unit of work

- Added `SchedsForums.Persistence.Services` to `Program.cs` to include persistence services.
- Modified `Program.cs` to add persistence services to the service collection and build the application.
- Removed `launchSettings.json` from the project.
- Updated `.gitignore` to include `FodyWeavers.xsd` and `SchedsForums.Api/Properties/launchSettings.json`.
- Changed `SchedsForums.Api.csproj` to include a project reference to `SchedsForums.Persistence`.
- Updated `ConfigureServices.cs` to return `IServiceCollection` from the `Configure` method.
- Modified `BaseEntity.cs` to make `CreatedAt` and `UpdatedAt` properties settable.
- Refactored `BaseUser.cs` to remove redundant code and improve readability.
- Updated `Admin.cs`, `Moderator.cs`, and `Student.cs` to include parameterless constructors.
- Reformatted `SchedsForums.Domain.csproj` and removed redundant project references.
- Added a project reference to `SchedsForums.Interface` in `SchedsForums.Infrastructure.csproj`.
- Refactored `PasswordService.cs` to implement `IBaseService`.
- Added `ForumsDbContext` class to manage database context and entity configurations.
- Created initial migration files for database schema.
- Updated `SchedsForums.Persistence.csproj` to include a project reference to `SchedsForums.Domain`.
- Added `DbContextRegistrationService.cs` to register the database context with dependency injection.

aabdoo23 committed:
repositories implemented

aabdoo23 committed:
repos fully done

Refactor namespaces and add repository services

Updated Program.cs to include new using directives and changed the namespace to SchedsForums.Api. Added calls to AddPersistenceServices and RegisterRepositories for DI. Refactored ConfigureServices.cs and ConfigureValidators.cs to use SchedsForums.Domain.ServiceConfiguration namespace. Updated SchedsForums.Infrastructure.csproj to reference SchedsForums.Domain. Added RepositoryRegistration.cs for repository service registration. Introduced new repository classes for Admin, Major, Moderator, and Student entities.

aabdoo23 committed:
Fixed the CLEAN Arch referencing issue, could use some refactoring later to ensure that everything in the right place

Did some work in the insert command for student section, but not done yet

Refactor namespaces and update project references

- Updated `Program.cs` to register persistence services and repositories.
- Modified `SchedsForums.Api.csproj` to reference `SchedsForums.Infrastructure`.
- Added `MediatR` package and folders for commands and queries in `SchedsForums.Application.csproj`.
- Cleaned up `SchedsForums.Domain.csproj` by removing unnecessary references.
- Updated `BaseEntity.cs` to use `SchedsForums.Domain.Interfaces` namespace.
- Refactored `BaseUser.cs` to `SchedsForums.Domain.Entities.Common` and simplified password handling.
- Removed unused using directives in `Major.cs`.
- Simplified exception message in `EntityNotFoundException.cs`.
- Updated repositories to use `SchedsForums.Domain.Interfaces` namespace.
- Updated `RepositoryRegistration.cs` to use `SchedsForums.Domain.Interfaces.Repositories` namespace.
- Cleaned up and updated `SchedsForums.Infrastructure.csproj` to include migrations folder and reference `SchedsForums.Domain`.
- Updated `PasswordService.cs` to use `SchedsForums.Domain.Interfaces` namespace.
- Updated `IMajorRepository.cs` to use `SchedsForums.Domain.Interfaces.Repositories` namespace.
- Updated `SchedsForums.Interface.csproj` to reference `SchedsForums.Domain`.
- Configured `BaseUser` entity with discriminator in `ForumsDbContext.cs`.
- Updated `SchedsForums.Persistence.csproj` to reference `SchedsForums.Domain`.
- Updated `SchedsForums.sln` to use new project GUIDs and removed references to `SchedsForums.Interface` and `SchedsForums.Persistence`.
- Added new DTOs, commands, and handlers for student creation.
- Added new interfaces for entities, services, and repositories.
- Added new `ForumsDbContext` in `SchedsForums.Infrastructure.Contexts`.
- Added `DbContextRegistrationService` in `SchedsForums.Persistence.Services`.

aabdoo23 committed:
CreateStudentCommand done (Sign up)

Refactor services, add MediatR, and update student creation

- Added `using` statements and updated `namespace` in `Program.cs`.
- Registered new services: `RegisterInfrastructureRepositories`, `RegisterInfrastructureServices`, `AddMediatRServices`.
- Added `MediatR` package and project references in `SchedsForums.Api.csproj`.
- Changed `CreateStudentCommand` return type to `StudentDTO`.
- Updated `CreateStudentCommandHandler` to use `FluentValidation` and `IPasswordService`.
- Modified `CreateStudentDTO` to inherit from `StudentDTO`.
- Simplified constructors in `BaseEntity` and `BaseUser`.
- Removed empty class body in `Major.cs` and `RepositoryRegistration` class.
- Added `Console.WriteLine` for connection string in `DbContextRegistrationService`.
- Refactored `PasswordService` to implement `IPasswordService`.
- Added `StudentsController` for student creation.
- Added initial migration files for database schema.
- Added `RegisterServices`, `CreateStudentValidator`, and `ServicesRegistration` classes.
- Added `IPasswordService` interface.

MostafaHgz commented on SchedsForums.Api/Controllers/StudentController.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public StudentsController(IMediator mediator)
+        {
+            _mediator = mediator;
```
add null check 
`_mediator = mediator ?? throw new ArgumentNullAgreement(nameof(mediator));`

MostafaHgz commented on SchedsForums.Api/Program.cs:
```diff
@@ -7,13 +11,19 @@ public static void Main(string[] args)
             var builder = WebApplication.CreateBuilder(args);
 
-            // Add services to the container.
 
```
remove extra line

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        private readonly IStudentRepository _studentRepository;
+        private readonly IPasswordService _passwordService;
+        private readonly IValidator<CreateStudentDTO> _validator = new CreateStudentValidator();
```
never use new in .net. we mostly use dependency injection for constructing any new object.

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public async Task<StudentDTO> Handle(CreateStudentCommand request, CancellationToken cancellationToken)
+        {
+            var validationResult = await _validator.ValidateAsync(request.Student);
```
from 24-28 will be always repeated in all commands/queries. and we need to keep the code dry
look up `AddValidatorFromAssembly()` in fluent validation docs

MostafaHgz commented on SchedsForums.Application/RegisterServices.cs:
```diff
@@ -0,0 +1,17 @@
+            services.AddMediatR(options => options.RegisterServicesFromAssemblies(typeof(CreateStudentCommandHandler).Assembly));
+
+            return services;
+
```
remove extra line

MostafaHgz commented on SchedsForums.Domain/Entities/Common/BaseUser.cs:
```diff
@@ -0,0 +1,21 @@
+        protected BaseUser()
+        {
+        }
+
```
remove extra line

MostafaHgz commented on SchedsForums.Domain/Interfaces/IGenericRepository.cs:
```diff
@@ -0,0 +1,12 @@
+        Task<T> InsertAsync(T entity);
+        Task<T> UpdateAsync(T entity);
+        Task<T> DeleteAsync(string id);
+
```
remove extra line

MostafaHgz replied:
also all interfaces should be in persistance. if you can't do it due to project dependencies we can check it when we next meet

aabdoo23 replied:
ok

MostafaHgz commented on SchedsForums.Domain/ServiceConfiguration/ConfigureServices.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public static IServiceCollection Configure(this IServiceCollection services)
+        {
+
```
remove extra line

MostafaHgz commented on SchedsForums.Domain/ServiceConfiguration/ConfigureServices.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public static class ConfigureServices
+    {
+        public static IServiceCollection Configure(this IServiceCollection services)
```
what does this do?

MostafaHgz commented on SchedsForums.Domain/ServiceConfiguration/ConfigureValidators.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public static class ConfigureValidators
+    {
+        public static IServiceCollection Configure(this IServiceCollection services)
```
you now have 2 extensions methods to IServiceCollection that are named the same. name the method itself a meaningful name
also this file does nothing

MostafaHgz commented on SchedsForums.Domain/Validators/BaseUserValidator.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    public class BaseUserValidator<T> : AbstractValidator<T> where T : BaseUser
+    {
+        public BaseUserValidator()
```
this validation is repeated in user validation as well. see how to use common validators in fluent validations

MostafaHgz commented on SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+        }
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.Entity<BaseUser>()
```
when you get to have 10 entities in the future this method will become 100 lines and will be a nightmare to read. 
check how to modularize model configuration and have a configuration file for each entity

MostafaHgz commented on SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+        public DbSet<Admin> Admins { get; set; }
+        public DbSet<Student> Students { get; set; }
+        public DbSet<Moderator> Moderators { get; set; }
+
```
class properties always come first before any methods or constructors
also 2 extra lines here

MostafaHgz commented on SchedsForums.Infrastructure/Repositories/AdminRepository.cs:
```diff
@@ -0,0 +1,16 @@
+        private readonly ForumsDbContext _context;
+        public AdminRepository(ForumsDbContext context) : base(context)
+        {
+            _context = context;
```
add null check for the context here

MostafaHgz commented on SchedsForums.Domain/Entities/Common/BaseUser.cs:
```diff
@@ -0,0 +1,21 @@
+            Email = email;
+            Password = password;
+        }
+        protected BaseUser()
```
remove all these empty constructors as they do nothing

MostafaHgz commented on SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public async Task<T> DeleteAsync(string id)
+        {
+            var entity = await _dbSet.FindAsync(id) ?? throw new NullReferenceException($"Can't find an entity with this Id: {id}.");
```
this is not null reference exception. this is when something that cannot be null 
also it's not the responsibility of the repository to throw errors. 
this method requires 2 query calls and it only needs one. see how you can delete with one query

MostafaHgz commented on SchedsForums.Infrastructure/Repositories/Common/GenericRepository.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public async Task<T> GetByIdAsync(string id)
+        {
+            return await _dbSet.FindAsync(id) ?? throw new NullReferenceException($"Can't find an entity with this Id: {id}.");
```
it's not the responsibility of the repository to throw the error. it should return a nullable value. use `SingleOrDefaultAsync`

aabdoo23 replied:
why singleOrDefault, im using findasync base on this:
![image](https://github.com/user-attachments/assets/f8ff78e4-9ccc-41a5-a48f-64417267c6ff)

MostafaHgz commented on SchedsForums.Infrastructure/Repositories/MajorRepository.cs:
```diff
@@ -0,0 +1,16 @@
+        private readonly ForumsDbContext _context;
+        public MajorRepository(ForumsDbContext context) : base(context)
+        {
+            _context = context;
```
remove all repositories that you didn't use in this story as they are not part of the story scope and you didn't use them
always add null checks for the when injecting. also you're injecting the ForumsDbContext inside the Major repository which is incorrect.

aabdoo23 replied:
I did remove the repo as it's not a part of the story ok.
but Im referencing the DbContext in the repo, why wrong though

MostafaHgz commented on SchedsForums.Infrastructure/Services/DbContextRegistrationService.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public static class DbContextRegistrationService
+    {
+        public static IServiceCollection AddPersistenceServices(this IServiceCollection services)
```
this adds the DB context not a service

MostafaHgz commented on SchedsForums.Infrastructure/Services/PasswordService.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace SchedsForums.Infrastructure.Services
+{
+    public class PasswordService : IPasswordService
```
what if I want to encrypt or hash any other thing than the password. this should be called `HashingService`

MostafaHgz commented on SchedsForums.Infrastructure/ServicesRegistration.cs:
```diff
@@ -0,0 +1,25 @@
+        {
+            services.AddScoped<IAdminRepository, AdminRepository>();
+            services.AddScoped<IModeratorRepository, ModeratorRepository>();
+            services.AddScoped<IStudentRepository, StudentRepository>();
```
only student repository is needed remove the rest

MostafaHgz commented on SchedsForums.Interface/IBaseService.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace SchedsForums.Interface
```
this does nothing

aabdoo23 replied:
yet

MostafaHgz commented on SchedsForums.Interface/Repositories/IAdminRepository.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Users;
```
remove all unwanted interfaces

MostafaHgz commented on SchedsForums.Interface/IGenericRepository.cs:
```diff
@@ -0,0 +1,12 @@
+        Task<T> InsertAsync(T entity);
+        Task<T> UpdateAsync(T entity);
+        Task<T> DeleteAsync(string id);
+
```
remove extra line

MostafaHgz commented on SchedsForums.Persistence/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+using SchedsForums.Domain.Entities.Users;
+namespace SchedsForums.Persistence.Contexts
+{
+    public class ForumsDbContext : DbContext
```
rename this to the name of the application

MostafaHgz replied:
you have this file twice

MostafaHgz commented on SchedsForums.Infrastructure/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    public class ForumsDbContext : DbContext
+    {
+        public ForumsDbContext(DbContextOptions<ForumsDbContext> options) : base(options)
```
you have this file twice

aabdoo23 replied:
i screwed with git, its not wallahy

MostafaHgz commented on SchedsForums.Infrastructure/Services/PasswordService.cs:
```diff
@@ -0,0 +1,18 @@
+        {
+            return BCrypt.Net.BCrypt.Verify(password, hash);
+        }
+
```
remove extra line

MostafaHgz commented on SchedsForums.Domain/Entities/Major.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Common;
```
only the user entity is related to the story's scope. remove all other entities

MostafaHgz commented on SchedsForums.Domain/Exceptions/EntityNotFoundException.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace SchedsForums.Domain.Exceptions
```
remove this exception and use .NET Exceptions

MostafaHgz commented on SchedsForums.Api/Controllers/StudentController.cs:
```diff
@@ -0,0 +1,28 @@
+
+            var result = await _mediator.Send(createStudentCommand);
+            return Ok(result);
+            //return CreatedAtAction(nameof(GetStudentById), new { id = result.Id }, result);
```
remove commented line

MostafaHgz commented on SchedsForums.Api/Controllers/StudentController.cs:
```diff
@@ -0,0 +1,28 @@
+        [HttpPost]
+        public async Task<IActionResult> CreateStudent([FromBody] CreateStudentDTO studentDto)
+        {
+            var createStudentCommand = new CreateStudentCommand { Student = studentDto };
```
you should make the body of the request is the command itself without constructing a new one

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentCommand.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class CreateStudentCommand : IRequest<StudentDTO>
+    {
+        public CreateStudentDTO Student { get; set; }
```
student dto should be the command. it doesn't make sense to have a field inside a class that has a class that has the data. just make the command have the data you want directly

MostafaHgz commented on SchedsForums.Application/BaseDTOs/StudentDTO.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace SchedsForums.Application.BaseDTOs
+{
+    public class StudentDTO
```
is this a request or response. you need to specify

aabdoo23 committed:
final updates based on PR

aabdoo23 committed:
Refactored the structure and re-referenced the whole project.

Refactor project structure and update repository usage

Updated CreateStudentCommandHandler to use new interfaces from SchedsForums.Application. Modified SchedsForums.Application.csproj to reference SchedsForums.Domain. Simplified BaseEntity by removing IBaseEntity implementation. Refactored BaseUser, Admin, Moderator, and Student to remove base class constructor calls. Removed obsolete interfaces from SchedsForums.Domain.Interfaces. Reintroduced OnModelCreating in SchedsForumsDbContext. Replaced GenericRepository with BaseRepository. Updated StudentRepository to inherit from BaseRepository and use new IStudentRepository interface. Updated SchedsForums.Infrastructure.csproj to reference SchedsForums.Application. Updated HashingService to use IHashingService interface. Added new interfaces and BaseRepository class to SchedsForums.Application.Interfaces and SchedsForums.Infrastructure.Repositories.Common respectively.
